pkg/client/client.go

```diff
@@ -167,98 +310,135 @@ func (c *ImmuClient) SetBatch(request *BatchRequest) (*schema.Index, error) {
 	start := time.Now()
 	if !c.isConnected() {
 		return nil, ErrNotConnected
 	}
-	result, err := c.serviceClient.Inclusion(context.Background(), &schema.Index{
+	result, err := c.serviceClient.Inclusion(ctx, &schema.Index{
 		Index: index,
 	})
 	c.Logger.Debugf("inclusion finished in %s", time.Since(start))
 	return result, err
 }
 
-func (c *ImmuClient) Consistency(index uint64) (*schema.ConsistencyProof, error) {
+func (c *ImmuClient) Consistency(ctx context.Context, index uint64) (*schema.ConsistencyProof, error) {
 	start := time.Now()
 	if !c.isConnected() {
 		return nil, ErrNotConnected
 	}
-	result, err := c.serviceClient.Consistency(context.Background(), &schema.Index{
+	result, err := c.serviceClient.Consistency(ctx, &schema.Index{
 		Index: index,
 	})
 	c.Logger.Debugf("consistency finished in %s", time.Since(start))
 	return result, err
 }
 
-func (c *ImmuClient) ByIndex(index uint64) (*schema.Item, error) {
+func (c *ImmuClient) ByIndex(ctx context.Context, index uint64) (*schema.Item, error) {
 	start := time.Now()
 	if !c.isConnected() {
 		return nil, ErrNotConnected
 	}
-	result, err := c.serviceClient.ByIndex(context.Background(), &schema.Index{
+	result, err := c.serviceClient.ByIndex(ctx, &schema.Index{
 		Index: index,
 	})
 	c.Logger.Debugf("by-index finished in %s", time.Since(start))
 	return result, err
 }
 
-func (c *ImmuClient) History(keyReader io.Reader) (*schema.ItemList, error) {
+func (c *ImmuClient) History(ctx context.Context, key []byte) (*schema.ItemList, error) {
 	start := time.Now()
 	if !c.isConnected() {
 		return nil, ErrNotConnected
 	}
-	key, err := ioutil.ReadAll(keyReader)
-	if err != nil {
-		return nil, err
-	}
-	result, err := c.serviceClient.History(context.Background(), &schema.Key{
+	result, err := c.serviceClient.History(ctx, &schema.Key{
 		Key: key,
 	})
 	c.Logger.Debugf("history finished in %s", time.Since(start))
 	return result, err
 }
 
-func (c *ImmuClient) Reference(keyReader io.Reader, valueReader io.Reader) (*schema.Index, error) {
+func (c *ImmuClient) Reference(ctx context.Context, reference []byte, key []byte) (*schema.Index, error) {
 	start := time.Now()
 	if !c.isConnected() {
 		return nil, ErrNotConnected
 	}
-	key, err := ioutil.ReadAll(valueReader)
-	if err != nil {
-		return nil, err
-	}
-	reference, err := ioutil.ReadAll(keyReader)
-	if err != nil {
-		return nil, err
-	}
-	result, err := c.serviceClient.Reference(context.Background(), &schema.ReferenceOptions{
+	result, err := c.serviceClient.Reference(ctx, &schema.ReferenceOptions{
 		Reference: &schema.Key{Key: reference},
 		Key:       &schema.Key{Key: key},
 	})
 	c.Logger.Debugf("reference finished in %s", time.Since(start))
 	return result, err
 }
 
-func (c *ImmuClient) ZAdd(setReader io.Reader, score float64, keyReader io.Reader) (*schema.Index, error) {
+// SafeReference ...
+func (c *ImmuClient) SafeReference(ctx context.Context, reference []byte, key []byte) (*VerifiedIndex, error) {
 	start := time.Now()
```

Comment:
All safe methods need to be thread safe.
Use mutex to lock immuclient properly, and release lock when methods return